rfcs/0000-top-level-assumptions.md

```diff
@@ -0,0 +1,327 @@
+This is implemented by [babel/babel#11689](https://github.com/babel/babel/pull/11689)).
+
+## Plugin API
+
+The first parameter passed to the plugins (often known as `api`) should have a new method: `assumptions(name: string): boolean | undefined`, which returns whether or not the assumption has been enabled. If a plugin is asking about an assumption not enabled or not supported by the used `@babel/core` version, it will return `undefined`.
+
+This is implemented as a function and not as an object whose properties reflect the assumptions because it configures the plugins' caching (they are reinstantiated when an assumption they use changes).
+
+## Assumptions list
+
+The different `loose` or `loose`-like options we currently have in our plugins map to these assumptions:
+
+> ℹ️ I'm not particularly attatched to these option names: I tried to choose something descriptive for their behavior, but they can all change.
+
+| Assumption | Behavior | Current option | Repl | Notes |
+|:-----------|:---------|:---------------|:----:|:-----:|
+| `ignoreToPrimitiveHint` | Transform `` `a${x}b` `` to `"a" + x + "b"` instead of `"a".concat(x, "b")` | `loose` in `transform-template-literals` |
+| `mutableTemplateObject` | Don't use `Object.freeze` for the template object created for tagged template literals. This effectively means using the `taggedTemplateLiteralLoose` helper instead of `taggedTemplateLiteral` | `loose` in `transform-template-literals` |
+| `ignoreFunctionLength` | The `.length` of a function should be cropped at the first default argument. When this option is enabled, ignore this spec requirement and don't rely on the `arguments` object | `loose` in `transform-parameters` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=GYVwdgxgLglg9mABMMAKAhgGkQI0QXkQCZsJsATAxAZgEpEBvAKAF8g&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) |
+| `iterableIsArray` | When using an iterable (in array destructuring, for-of or with spreads), assume that the iterable object is an `Array` | `loose` in `transform-destructuring` and `transform-spread`, `assumeArray` in `transform-for-of` | [🔗](https://babeljs.io/repl#?browsers=defaults&build=&builtIns=false&spec=false&loose=false&code_lz=G4QwTgBCELwQ2gOmQIwLoG4BQpLwB4A0EyiAnmrFNkA&debug=false&forceAllTransforms=false&shippedProposals=true&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=script&lineWrap=true&presets=env%2Cenv&prettier=false&targets=&version=7.10.1&externalPlugins=) (it doesn't support `assumeArray`) | 1 |
```

Comment:
`arrayLikeIsIterable` makes things like `{ 0: "a", 1: "b", length: 2 }` iterable, in addition to all the normal iterables. This is especially useful when working, for example, with DOM collections that _should_ be iterable but aren't in all the browsers, or with libraries that implement `Symbol.iterator` only when `Symbol` exists.

`iterableIsArray` is the opposite: rather than extending the definition of iterable, it restricts it. It ignores the iterator protocol, making it possible to produce much smaller code. For example, `for (let x of arr) {}` can be compiled to `for (let i = 0; i < arr.length; i++) { let x = arr[i] }`.

So the first one violates the spec by allowing more (for better compatibility), while the second one violates the spec by allowing less (for smaller generated code).

Reply:
Yes, it does.